### DISCUSSION TITLE
添加最小化到托盘功能后 双击图标无法弹出界面

### DISCUSSION THREAD

**Discussion by hczhouu** [netease-im/NIM_Duilib_Framework#454] [General]
使用showwindow(false, false)最小化到托盘后
在双击消息里面showwindow(true, true)无法弹出主界面  在任务栏有显示 但界面无法弹出
是我的使用方法不对吗

**Comment by nmgwddj** [comment]
参考 NIM Demo 示例：https://github.com/netease-kit/NIM_PC_Demo/blob/a3a504a33c957922edcdfdfbccbbbc36bc729cc4/nim_win_demo/gui/main/main_form.cpp#L68-L78

**Reply by hczhouu** [reply]
按照示例里面写的也是无法弹出   只在任务栏有显示

**Reply by nmgwddj** [reply]
将隐藏窗口的代码按这种方式试一下是否可行：
https://github.com/netease-kit/NIM_PC_Demo/blob/a3a504a33c957922edcdfdfbccbbbc36bc729cc4/nim_win_demo/gui/main/main_form_ui.cpp#L204-L208

**Reply by hczhouu** [reply]
就是按这个方式呀  无法弹出主界面  win10系统
![QQ截图20220915140739](https://user-images.githubusercontent.com/39211209/190327213-7a8a8d87-be98-4e09-85a1-6e9c2a9d353a.png)
![微信截图_20220915140807](https://user-images.githubusercontent.com/39211209/190327219-91e1f18d-7150-48cb-ab87-dae729a0e87e.png)

**Reply by nmgwddj** [reply]
将您的实现代码推送到自己的仓库中，提供一个链接我们一起参考研究一下。

**Reply by hczhouu** [reply]
https://github.com/AidanMichle/wgupdate  帮忙看下

**Reply by hczhouu** [reply]
这个有解决办法吗

**Comment by nmgwddj** [comment]
重命名 xml 中的关闭按钮名称（duilib 内部有 btnclose 的默认实现会导致该问题）如 `btn_close`，然后 attch 重命名后的关闭按钮点击事件，实现最小化和隐藏功能，示例：

```cpp
void BasicForm::InitWindow()
{
	auto close = dynamic_cast<ui::Button*>(FindControl(L"btn_close"));
	if (close) {
		close->AttachClick([this](ui::EventArgs* event) {
			SendMessage(WM_SYSCOMMAND, SC_MINIMIZE);
			::ShowWindow(m_hWnd, SW_HIDE);
			nbase::ThreadManager::PostDelayedTask(0, [this]() {
				this->ActiveWindow();
				::SetForegroundWindow(m_hWnd);
				::BringWindowToTop(m_hWnd);
			}, nbase::TimeDelta::FromSeconds(2));
			return true;
		});
	}
}

LRESULT BasicForm::OnClose(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
{
	return __super::OnClose(uMsg, wParam, lParam, bHandled);
}
```

**Comment by hczhouu** [comment]
OK, 已解决
